### PR TITLE
chore: move RouterData Request types to hyperswitch_domain_models crate

### PR DESCRIPTION
**Pull request by hrithikesh026**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Moved RouterData::Request types to hyperswitch domain models crate.

Other changes:
Moved `enum ApiErrorResponse` to `hyperswitch domain models` as well

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Checked if the code compiles successfully.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by jarnura — DISMISSED** [review]

File: crates/common_utils/src/custom_serde.rs [hunk 1]
```diff
@@ -266,6 +266,15 @@ pub mod iso8601custom {
...
 }
 
+/// serde Serialize using display triat implementation
+pub fn display_serialize<T, S>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
+where
+    T: std::fmt::Display,
+    S: serde::ser::Serializer,
+{
+    serializer.serialize_str(&format!("{}", value))
+}
```
Comment (SanchithHegde) [inline crates/common_utils/src/custom_serde.rs:276]: Nit: Spelling: `triat` -> `trait`

Regardless, if the need today is only within the `router` crate, we can consider using [`serde_with::DisplayFromStr`](https://docs.rs/serde_with/3.8.1/serde_with/struct.DisplayFromStr.html) instead of writing an implementation ourselves, and since we pull the dependency already. The example they have includes `mime::Mime` itself!

File: crates/hyperswitch_domain_models/Cargo.toml [hunk 2]
```diff
@@ -14,11 +14,17 @@ payouts = ["api_models/payouts"]
...
 common_enums = { version = "0.1.0", path = "../common_enums" }
 common_utils = { version = "0.1.0", path = "../common_utils" }
 masking = { version = "0.1.0", path = "../masking" }
 diesel_models = { version = "0.1.0", path = "../diesel_models", features = ["kv_store"] }
+cards = {version = "0.1.0", path = "../cards"}
+router_derive = {version = "0.1.0", path = "../router_derive"}
+mime = "0.3.17"
+url = { version = "2.5.0", features = ["serde"] }
+utoipa = { version = "4.2.0", features = ["preserve_order", "preserve_path_order", "time"] }
+serde_with = "3.7.0"
```
Comment (SanchithHegde) [inline crates/hyperswitch_domain_models/Cargo.toml:27]: Please segregate (and later, sort) them as first and third party dependencies correctly: `mime`, `serde_with`, `url` and `utoipa` are third party depndencies.

File: crates/hyperswitch_domain_models/src/errors/api_error_response.rs [hunk 3]
```diff
@@ -0,0 +1,637 @@
+#![allow(dead_code, unused_variables)]
```
Comment (SanchithHegde) [inline crates/hyperswitch_domain_models/src/errors/api_error_response.rs:1]: Can you check if these two warnings need to be allowed, or if we can remove this annotation?

File: crates/hyperswitch_domain_models/src/errors/api_error_response.rs [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+    ProcessingError,
+    BadGateway,
+    ServerNotAvailable,
+    DuplicateRequest,
+    ValidationError,
+    ConnectorError,
+    LockTimeout,
+}
+
+#[allow(dead_code)]
```
Comment (SanchithHegde) [inline crates/hyperswitch_domain_models/src/errors/api_error_response.rs:24]: Same here, do we need this annotation?

File: crates/router/src/compatibility/stripe/errors.rs [hunk 5]
```diff
@@ -1,7 +1,8 @@
 #![allow(unused_variables)]
```
Comment (SanchithHegde) [inline crates/router/src/compatibility/stripe/errors.rs:1]: Unrelated to this PR, but do we need this annotation?

File: crates/router/src/core/errors.rs [hunk 6]
```diff
@@ -11,15 +10,16 @@ use std::fmt::Display;
 use actix_web::{body::BoxBody, ResponseError};
 pub use common_utils::errors::{CustomResult, ParsingError, ValidationError};
 use diesel_models::errors as storage_errors;
-pub use hyperswitch_domain_models::errors::StorageError as DataStorageError;
+pub use hyperswitch_domain_models::errors::{
+    api_error_response::*, StorageError as DataStorageError,
```
Comment (SanchithHegde) [inline crates/router/src/core/errors.rs:14]: Can we avoid wildcard imports, and instead import required items?

File: crates/router/src/services/api.rs [hunk 7]
```diff
@@ -1328,6 +1328,11 @@ impl EmbedError for Report<api_models::errors::types::ApiErrorResponse> {
     }
 }
 
+impl EmbedError
```
Comment (SanchithHegde) [inline crates/router/src/services/api.rs:1331]: I think we can remove this entire trait, we don't use it anywhere at all, from what I know.
This can be taken up in a separate PR, please create an issue and mark it as a good first issue.

File: crates/router/src/utils/custom_serde.rs
Comment (SanchithHegde) [inline crates/router/src/utils/custom_serde.rs:1]: Remove this file and the `mod` declaration here:
https://github.com/juspay/hyperswitch/blob/247a171abfd8a523e4074771dbe615be852f6b12/crates/router/src/utils.rs#L4

File: crates/router/src/types/domain/payments.rs [hunk 8]
```diff
@@ -1,837 +1 @@
...
-            }
-            api_models::payments::BankTransferData::Pix {} => Self::Pix {},
-            api_models::payments::BankTransferData::Pse {} => Self::Pse {},
-            api_models::payments::BankTransferData::LocalBankTransfer { bank_code } => {
-                Self::LocalBankTransfer { bank_code }
-            }
-        }
-    }
-}
+pub use hyperswitch_domain_models::payment_method_data::*;
```
Comment (SanchithHegde) [inline crates/router/src/types/domain/payments.rs:1]: If it isn't asking for too much, can we avoid the wildcard import here?

**Review by SanchithHegde — APPROVED** [review]

**Review by jarnura — APPROVED** [review]
